clients/src/main/resources/common/message/FetchRequest.json

```diff
@@ -50,14 +50,23 @@
   // Version 13 replaces topic names with topic IDs (KIP-516). May return UNKNOWN_TOPIC_ID error code.
   //
   // Version 14 is the same as version 13 but it also receives a new error called OffsetMovedToTieredStorageException(KIP-405)
-  "validVersions": "0-14",
+  //
+  // Version 15 adds the ReplicaState which includes new field ReplicaEpoch and the ReplicaId. Also,
+  // deprecate the old ReplicaId field and set its default value to -1. (KIP-903)
+  "validVersions": "0-15",
   "flexibleVersions": "12+",
   "fields": [
     { "name": "ClusterId", "type": "string", "versions": "12+", "nullableVersions": "12+", "default": "null",
       "taggedVersions": "12+", "tag": 0, "ignorable": true,
       "about": "The clusterId if known. This is used to validate metadata fetches prior to broker registration." },
-    { "name": "ReplicaId", "type": "int32", "versions": "0+", "entityType": "brokerId",
+    { "name": "ReplicaId", "type": "int32", "versions": "0-14", "default": "-1", "entityType": "brokerId",
       "about": "The broker ID of the follower, of -1 if this request is from a consumer." },
+    { "name": "ReplicaState", "type": "ReplicaState", "taggedVersions":"15+", "tag": 1, "fields": [
```

Comment:
Sorry if I missed this conversation, but why is this tagged? Is it to avoid the IBP bump? We would still need it if we no longer set replica ID right?

Reply:
It is to reduce the message size for consumer cases where the replicaId and replicaEpoch are both -1. When the values are default values (both fields are -1), the replicaState will not be parsed in the message.

Reply:
I see. So we still rely on IBP to know the version?

Reply:
Also, does the builder still populate the field? Does it automatically remove it if we keep the defaults?

Reply:
> So we still rely on IBP to know the version?

Yes
> does the builder still populate the field?

Yes, when constructing an instance, it will do this.replicaState = new ReplicaState()
> Does it automatically remove it if we keep the defaults

When serialize the request:
```
if (!this.replicaState.equals(new ReplicaState())) {
                    _writable.writeUnsignedVarint(1);
                    _writable.writeUnsignedVarint(this.replicaState.size(_cache, _version));
                    replicaState.write(_writable, _cache, _version);
 }
```

Reply:
> Yes, when constructing an instance, it will do this.replicaState = new ReplicaState()

The builder is still going to write to these feels as far as I can see in the code. (FetchRequest line 266) Its just the case that these happen to be the defaults and so the state is equal, right?

Reply:
Correct.

Reply:
The tagged field magic happens in the lower layer of the FetchRequestData.java, which is auto-generated.